### PR TITLE
Weights speedup

### LINKED ISSUE #16: Docs Build Hard-Set to Python 3.8
Once the matplotlib dependencies support Python 3.9, we should reset our docs build to use the latest version of Python (in the workflow yaml this would be `'3.x'` as opposed to `'3.8'`). 

See [here](https://github.com/nucleic/kiwi/issues/92) for the upstream issue that once closed, should allow us to move back to `'3.x'`.

### PR DESCRIPTION
**Pull request by elbeejay**
### speeding up the random walk
This PR takes a suggestion from @wrightky, and separates the routing weight creation from the actual particle random walking itself. Actual use of the code does not change at all from the perspective of someone running functions in `routines.py` or calling `run_iteration()` from `particle_track.py` (which is good). 

Internally this PR split `get_weight()` into `make_weight()` and `get_weight()` in `lagrangian_walker.py`. 
- `make_weight()` is called when the `Particles` class is initialized. All of the routing weights are computed at that time and stored in a L x W x 9 array (where L and W are the length and width of the domain). A progress bar was added for this initialization as it may take some time on larger domains. 
- `get_weight()` is now a pretty small function that just calls either the random picker or steepest descent function, depending on the parameters of the run

Example scripts seem to run at about the same speed as they used to, as most of their processing time gets tied up in plot creation and saving. But I think that for people using the `run_iteration()` functions to simulate many particles, this change will greatly speed things up. 

**Note:** Unit tests of example cases and checks on individual particle stepping all pass because while the weighting calculation is changed, the "random selection" process for the particle walking hasn't changed - meaning that previous tests which set up the rng-seed to create deterministic (reproducible) results are still working! This also implies that any previous simulations conducted with `dorado` will be identical to new ones, the new simulations will just run faster.

This PR also closes #16 by allowing the documentation to be built on Python 3.9. This implies that `dorado` will work on Python 3.9, but as we don't explicitly run unit tests on Python 3.9 I don't know that we should say we 'officially' support it ... yet.

### Speed comparisons
Comparing the speed of particle routing for code in this PR vs current code. All conditions are steady flow fields, and particles are stepped by iterating over `run_iteration()` a bunch of times (no routines are used), each condition in the table below is run 3 times and an approximate average time is reported. These are tests run on the example domains. Also worth mentioning is that these time tests were run on Python 3.8.3.

| Domain Used | Number of Particles | Number of Iterations | Current Speed [s] | This PR [s] |
| :-: | :-: | :-: | :-: | :-: |
| anuga example domain | 100 | 50 | 0.8 | 1.3 |
| anuga example domain | 1000 | 50 | 8.0 | 2.6 |
| anuga example domain | 5000 | 50 | 41.0 | 8.1 |
| deltarcm example domain | 100 | 50 | 0.7 | 3.5 |
| deltarcm example domain | 1000 | 50 | 7.5 | 4.8 |
| deltarcm example domain | 5000 | 50 | 38.0 | 10.2 |

The new method of computing weights a priori costs some time upfront compared to the current method of computing weights on the fly. This makes simulations with few particles slower with the new method. Once the number of particles becomes large enough however, the cost of computing weights on the fly catches up to the a priori method - on the example domains the method in this PR is 3-5x faster than the current implementation when 5,000 particles are simulated.

### outstanding items to complete before PR is 'ready'
- [x] additional testing to ensure weights are being generated as expected
- [x] more tests for "edge" cases in both senses of the word (are boundaries acting as expected?)
- [x] added test to ensure that "overflow" scenario in 'exact' particle generation works (e.g. when number of particles exceeds seed locations the assignment loops through seed locations until all particles are generated)
- [x] added test for scenarios where `walk_data` is defined but the `Particles` object does not have explicit information about `Np_tracer` (number of particles). test to ensure that this is properly counted when `run_iteration()` is used to iterate the particles in `walk_data`
- [x] some actual time tests against current code to better gauge speed-up
- [x] check parallel code to ensure weights only get calculated once there (to ensure it is faster than serial code) - yes, one of the inputs to the `parallel_routing()` function is a pre-initialized `Particles` class
- [x] make any necessary changes to documentation to cover this procedural change (weight computation a priori vs for each step)
- [x] increment version number in `__init__.py`

### CODE REVIEW

File: dorado/lagrangian_walker.py [hunk 1]
```diff
@@ -38,11 +39,73 @@ def random_pick_seed(choices, probs=None):
...
 
+def make_weight(Particles):
+    """Make an array with the routing weights."""
+    # init the weight array
+    L, W = np.shape(Particles.stage)
+    Particles.weight = np.zeros((L, W, 9))
+    # do weighting calculation for each cell
+    print('Calculating routing weights ...')
+    for i in tqdm(list(range(1, L-1)), ascii=True):
+        for j in list(range(1, W-1)):
```
Comment (wrightky) [inline dorado/lagrangian_walker.py:50]: So, looking at this code, I see that the main structure of the weight computation hasn't changed. We're still constructing small sub-arrays at each index, doing a few quick operations (max, cleaning, multiplications, summing), and saving the 9 resulting weights in a `weight` array. Surprised we hadn't tried this sooner given how much better it performs.

It looks like the key reason the runtime scales better in this model isn't anything about how this computation works, it's how many times we do it. Originally, we performed these operations locally once for every particle at every iteration. So, the runtime scaled as `Np_tracer * iterations_per_particle`. Now, we perform this once for every cell, so it scales with domain size `(L-2) * (W-2)`. I bet if you checked the example cases you benchmarked, the ratio of these values should give roughly the speedup you observed. 

One thing I wonder, though, is whether we could obtain even faster runtimes by modifying the structure of this computation itself, by switching from local array operations to global. Whatever is causing this function to take so long must be due to the fact that we're repeatedly constructing many sub-arrays in a loop and operating on them, instead of performing a few big global array operations (inside which we'd be taking advantage of all the fast numpy broadcasting stuff). In principal, if we broke up these operations (max, cleaning, multiplication, summing) into a loop over each of the D8 directions, with each being a global matrix operation, instead of a loop over each cell, we could reduce the amount of overhead repeatedly calling these functions. I don't know exactly how that scales, but it'd be the difference between `time(np.max(small array)) * (L-2) * (W-2)` and `time(np.max(big array)) * 9`. Does that scale better? Not sure.

That being said, I think it's very fine to commit these changes now -- the PR looks good, and it's still a significant improvement in runtime. Might be worth investigating if the global method will be better in a future change.

File: dorado/lagrangian_walker.py [hunk 2]
```diff
@@ -38,11 +39,73 @@ def random_pick_seed(choices, probs=None):
...
+            # weights for each location in domain
+            # get stage values for neighboring cells
+            stage_ind = Particles.stage[i-1:i+2, j-1:j+2]
+
+            # calculate surface slope weights
+            weight_sfc = np.maximum(0,
```
Comment (wrightky) [inline dorado/lagrangian_walker.py:56]: You should switch to an explicit import (`from numpy import maximum`) to shave off a bit more time

Comment (wrightky) [reply dorado/lagrangian_walker.py:56]: Probably also `nansum`, maybe `zeros_like`, anything we're doing here a bunch of times

File: dorado/lagrangian_walker.py (hunk 1 above)
Comment (elbeejay) [reply dorado/lagrangian_walker.py:50]: Yeah this is a good point. 

It is pretty much just a port of the previous `get_weight()` functionality to doing it upfront for all locations. I agree that the trade-off in time with the methods should just about match what you are saying. 

I'm going to open up an issue with your commentary on this so that we don't forget to implement it in the future.

File: dorado/lagrangian_walker.py (hunk 2 above)
Comment (elbeejay) [reply dorado/lagrangian_walker.py:56]: I am not convinced that importing in this style is actually faster. I couldn't find anything online suggesting that it is the case? 

Ran some tests that basically ran code that looked like this (e.g. with `maximum()` function):
```
import time
import numpy as np
from numpy import maximum

t0 = time.time()
for i in range(500):
    np.maximum(arr, 1)
print('np time: ' + str(time.time() - t0))

t0 = time.time()
for i in range(500):
    maximum(arr, 1)
print('func time: ' + str(time.time() - t0))
```

And got results that were super close in times with the `np.maximum()` version sometimes fast and sometimes slower than the `maximum()` version.

Comment (wrightky) [reply dorado/lagrangian_walker.py:56]: I learned this a few months ago when looking into optimizing something in `run_iteration`. People on stack said it can matter if you're looking up a function many thousands of times, which given the size of our domains seems feasible. Do you see a time difference if you change 500 to like 1e6? 
https://stackoverflow.com/questions/32151193/is-there-a-performance-cost-putting-python-imports-inside-functions

Comment (elbeejay) [reply dorado/lagrangian_walker.py]: It wasn't a big deal so I just imported the commonly called numpy functions within the function. 

For tests I did on 10,000 calls, the differences in time were <0.5s, and still flip-flopped a bit (maybe due to other CPU overhead at the time), but the change is simple and I don't think it makes the function any more or less confusing.

**Comment by wrightky** [comment]
Everything looks good! Going to merge this now.
